[Security] Deprecate remaining anonymous checks

wouterj opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | Ref #41613
| License       | MIT
| Doc PR        | tbd

Deprecates the remaining checks for anonymous found in #41613.  It's WIP because the tests are failing until #42423 is merged and this PR is rebased (didn't update one test to avoid merge conflicts).

Besides this, it also introduced `IS_AUTHENTICATED` and `AuthenticationTrustResolver::isAutenticated()`. Previously, `IS_AUTHENTICATED_ANONYMOUSLY` was considered to be the "bottom type" for authenticated requests. As this is no longer true, `IS_AUTHENTICATED_REMEMBERME` is now the new "bottom type". I suggest we use an explicit bottom type (the ones introduced) instead to avoid another such update if we change something with remember me. It's also more clear on the exact intent of the check.

wouterj committed:
[Security] Deprecate remaining anonymous checks

Nyholm commented on src/Symfony/Component/Security/Core/Authentication/AuthenticationTrustResolverInterface.php:
```diff
@@ -27,6 +29,8 @@ interface AuthenticationTrustResolverInterface
      *
      * @return bool
+     *
+     * @deprecated since 5.4
```
Should we say: 
```suggestion
     * @deprecated since 5.4, use isAuthenticated() instead.
```

Nyholm reviewed (commented):
Thank you. Well done

After first review, It looks good. Just waiting for it to remove WIP.

wouterj commented:
Thanks for the reviews @Nyholm!

> Just waiting for it to remove WIP

The WIP is just because of the intentionally broken tests. Everything except `ExpressionLanguageTest` is in the final state as far as I'm concerned.

chalasr commented:
The most common usage of `IS_AUTHENTICATED_ANONYMOUSLY` looks like:

```yaml
firewalls:
    main:
        pattern: ^/
        form_login:
            login_path: login
            check_path: login

access_control:
    - { path: ^/login, roles: IS_AUTHENTICATED_ANONYMOUSLY }
    - { path: ^/another_public_resource, roles: IS_AUTHENTICATED_ANONYMOUSLY }
    - { path: ^/, roles: IS_AUTHENTICATED_FULLY }
```

What is the new way to achieve that? How should one allow a single route to be accessed without being authenticated?
I think having such before/after example in the UPGRADE file would be useful.

fabpot commented:
#42423 has been merged now.

wouterj commented:
Good idea @chalasr. The great thing: this made me realize that I was suggesting a wrong replacement. I've added your example to the UPGRADE guide.

Meanwhile, I also fixed the tests and updated the `isAuthenticated()` check to take into account deprecated features (and also make sure to be forward compatible by returning `false` for anonymous).

---

I think I have to disagree with fabbot here, their suggestion doesn't make sense: (I know we like to use the pixels at the right side of the screen, but I'm going to need a much much wider screen to be able to read this line :smile: )

```diff
diff -ru src/Symfony/Component/Security/Http/Tests/Firewall/ContextListenerTest.php src/Symfony/Component/Security/Http/Tests/Firewall/ContextListenerTest.php
--- src/Symfony/Component/Security/Http/Tests/Firewall/ContextListenerTest.php	2021-08-14 09:17:15.170647398 +0000
+++ src/Symfony/Component/Security/Http/Tests/Firewall/ContextListenerTest.php	2021-08-14 09:17:18.381551885 +0000
@@ -416,11 +416,9 @@
 
         $tokenStorage = new TokenStorage();
         $usageIndex = $session->getUsageIndex();
-        $tokenStorage = new UsageTrackingTokenStorage($tokenStorage, new class(
-            (new \ReflectionClass(UsageTrackingTokenStorage::class))->hasMethod('getSession') ? [
-                'request_stack' => function () use ($requestStack) {
-                return $requestStack;
-            }] : [
+        $tokenStorage = new UsageTrackingTokenStorage($tokenStorage, new class((new \ReflectionClass(UsageTrackingTokenStorage::class))->hasMethod('getSession') ? ['request_stack' => function () use ($requestStack) {
+            return $requestStack;
+        }] : [
                 // BC for symfony/framework-bundle < 5.3
                 'session' => function () use ($session) {
                     return $session;
```

chalasr reviewed (approved):
👍

fabpot commented:
Thank you @wouterj.
